src/passes/Vacuum.cpp

```diff
@@ -49,155 +50,85 @@ struct Vacuum : public WalkerPass<ExpressionStackWalker<Vacuum>> {
     walk(func->body);
   }
 
-  // Returns nullptr if curr is dead, curr if it must stay as is, or another
-  // node if it can be replaced. Takes into account:
+  // Returns nullptr if curr is dead, curr if it must stay as is, or one of its
+  // children if it can be replaced. Takes into account:
+  //
   //  * The result may be used or unused.
-  //  * The type may or may not matter (a drop can drop anything, for example).
+  //  * The type may or may not matter.
+  //
+  // For example,
+  //
+  //  (drop
+  //   (i32.eqz
+  //    (call ..)))
+  //
+  // The drop means that the value is not used later. And while the call has
+  // side effects, the i32.eqz does not. So when we are called on the i32.eqz,
+  // and told the result does not matter, we can return the call. Note that in
+  // this case the type does not matter either, as drop doesn't care, and anyhow
+  // i32.eqz returns the same type as it receives. But for an expression that
+  // returns a different type, if the type matter then we cannot replace it.
   Expression* optimize(Expression* curr, bool resultUsed, bool typeMatters) {
     FeatureSet features = getModule()->features;
     auto type = curr->type;
-    // An unreachable node must not be changed.
+    // If the type is none, then we can never replace it with another type.
+    if (type == Type::none) {
+      typeMatters = true;
+    }
+    // An unreachable node must not be changed. DCE will remove those.
     if (type == Type::unreachable) {
       return curr;
     }
-    // We iterate on possible replacements. If a replacement changes the type,
-    // stop and go back.
+    // resultUsed only makes sense when the type is concrete
+    assert(!(resultUsed && curr->type == Type::none));
```

Comment:
Nit: I think the comment makes the intention clear so that this is more readable.
```suggestion
    assert(!resultUsed || curr->type != Type::none));
```

Reply:
That's funny, for me it's the opposite... :smile: but let's go with your way, to favor people reading the code.